File: sorting_algorithms.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>


// =-=-=-= Insertion Sort  =-=-=-= //

void insertion(int vtr[],int tmnh)
{
	for (int i = 1; i < tmnh; i++) {
		int temp = vtr[i], j = i;
		while (j > 0 && vtr[j-1] > temp) {
			vtr[j] = vtr[j - 1];
			j = j - 1;
		}
		vtr[j] = temp;
	}
}

// =-=-=-= Selection Sort  =-=-=-= //


void troca(int vtr[], int antigo, int novo)
{
	int aux;
	aux = vtr[antigo];
	vtr[antigo] = vtr[novo];
	vtr[novo] = aux;

}

void selection(int vtr[], int tmnh)
{
	int indxMnr, aux;
	for (int i = 0; i < tmnh -1; i++) {
		indxMnr = i;
		for (int j = i+1;j < tmnh; j++) {
			if (vtr[j] < vtr[indxMnr]) {
				indxMnr = j;
				troca(vtr,i,j);
			}
		}
	}
}






// =-=-=-= Bubble Sort  =-=-=-= //

void bubble(int vtr[],int tmnh){
	int aux = 0;
	for (int i = 0; i < tmnh; i++) {
		for (int j = 0; j < (tmnh - i - 1); j++) {

			if (vtr[j] > vtr[j+1]) {
				aux = vtr[j];
				vtr[j] = vtr[j+1];
				vtr[j+1] = aux;

			}
		}
	}
}

// =-=-=-= Heap Sort  =-=-=-= //

void fazerHeap(int vtr[], int tmnh, int iAtual) {
	int esq, drta, pai, aux ;
	pai = iAtual;
	esq = (2*iAtual) ;
	drta = (2*iAtual) + 1;
	if(esq < tmnh && vtr[esq] > vtr[pai]) {
		pai = esq;

	} 
	if(drta < tmnh && vtr[drta] > vtr[pai]){
		pai = drta;
	}
	if(pai != iAtual){
		aux = vtr[iAtual];
		vtr[iAtual] = vtr[pai];
		vtr[pai] = aux;
		fazerHeap(vtr,tmnh,pai);
	}
}    

void heapSort(int vtr[], int tmnh) {
	int aux, tmnhAux;
	int i = (tmnh/2) - 1;
	tmnhAux = tmnh -1;
	
	for(i; i >= 0; i--){
		fazerHeap(vtr,tmnh,i);
	}
	for(tmnhAux; tmnhAux >= 0; tmnhAux--){
		aux = vtr[tmnhAux];
		vtr[tmnhAux] = vtr[0];
		vtr[0] = aux;
		fazerHeap(vtr,tmnhAux,0);

	}
}

// =-=-=-= Merge Sort =-=-=-= //

void merge(int vtr[],int esq, int met, int dir){
	int tmnhEsq, tmnhDrta, aux;
	int idxAuxEsq, idxAuxDrta;
	tmnhEsq = met - esq + 1;
	tmnhDrta = dir - met;
	idxAuxEsq = 0;
	idxAuxDrta = 0;
	aux = esq;
	int vtrEsq[tmnhEsq];
	int vtrDrta[tmnhDrta];

	for(int i = 0; i < tmnhEsq; i++){
		vtrEsq[i] = vtr[i+esq];
	}
	for(int j = 0; j < tmnhDrta; j++){
		vtrDrta[j] = vtr[met+1+j];

	}


	while((idxAuxEsq < tmnhEsq ) && (idxAuxDrta < tmnhDrta)){

		if(vtrEsq[idxAuxEsq] <= vtrDrta[idxAuxDrta]){
			vtr[aux] = vtrEsq[idxAuxEsq];
			idxAuxEsq++;
		} else {
			vtr[aux] = vtrDrta[idxAuxDrta];
			idxAuxDrta++;
		}
		aux++;
	}
	while(idxAuxEsq < tmnhEsq){
	  vtr[aux] = vtrEsq[idxAuxEsq];
	  idxAuxEsq++;
	  aux++;
	}

	while(idxAuxDrta < tmnhDrta){
	  vtr[aux] = vtrDrta[idxAuxDrta];
	  idxAuxDrta++;
	  aux++;
	}
  }


void mergeRecursiva(int vtr[], int idxEsq, int idxDrta ){
	if(idxEsq < idxDrta){
		int idxMetade;
		idxMetade = idxEsq + (idxDrta - idxEsq)/2;

		mergeRecursiva(vtr,idxEsq,idxMetade);
		mergeRecursiva(vtr,idxMetade+1,idxDrta);
		merge(vtr,idxEsq,idxMetade,idxDrta);
	}
}


void mergeSort(int vtr[], int tmnh){
	
	mergeRecursiva(vtr, 0, tmnh - 1);

}

// =-=-=-= Quick Sort =-=-=-= //

// troca = mesmo do selection

int particiona(int vtr[], int esq, int drta) {	
	int i = esq, j = drta ;
	int pivo;
	pivo = vtr[esq];

	while(i < j){
		while(vtr[i] <= pivo){
			i++;
		}
		while(vtr[j] > pivo){
			j--;
		}
		if(i < j){
			troca(vtr,i,j);
		}
	}
	troca(vtr,esq,j);
	return j;

}

void quickRecursiva(int vtr[], int esq, int drta) {
	int r =  0;
	if(drta >= esq) {
		r = particiona(vtr, esq, drta);
		quickRecursiva(vtr, esq, r-1);
		quickRecursiva(vtr, r+1, drta);
	}
}
void quickSort(int vtr[], int tmnh){
	
	quickRecursiva(vtr, 0, tmnh - 1);

}

// =-=-=-= Print Array =-=-=-= //

void imprimir(int vtr[], int tmnh) {

	for(int i = 0; i < tmnh ; i++){
			printf("%d ",vtr[i]);
	}

}   

// =-=-= Selecionar Arquivo =-=-= //

void selecionar_arquivo(int array[], int *tamanho) {
	
	printf("\n");
	printf("Selecione o arquivo que você deseja executar: \n");
	printf("1) 1000                                       \n");
	printf("2) 5000                                       \n");
	printf("3) 10000                                      \n");
	printf("4) 20000                                      \n");
	printf("5) 50000                                      \n");
	printf("6) 75000                                      \n");
	printf("7) 100000                                     \n");

	int arq;
	scanf("%d", &arq);
	FILE *arquivo;

	if (arq == 1) {

		arquivo = fopen("1000.txt", "r");
		*tamanho = 1000;

	} else if (arq == 2) {
		
		arquivo = fopen("5000.txt", "r");
		*tamanho = 5000;

	} else if (arq == 3) {

		arquivo = fopen("10000.txt", "r");
		*tamanho = 10000;
			
	} else if (arq == 4) {

		arquivo = fopen("20000.txt", "r");
		*tamanho = 20000;

	} else if (arq == 5){
	
		arquivo = fopen("50000.txt", "r");
		*tamanho = 50000;

	} else if (arq == 6) {
		
		arquivo = fopen("75000.txt", "r");
		*tamanho = 75000;

	} else if (arq == 7) {
		
		arquivo = fopen("100000.txt", "r");
		*tamanho = 100000;
	
	} else {
		
		printf("Valor inválido!");
	
	}

	int cont = 0;
	while (!feof(arquivo)) {

		fscanf(arquivo,"%d",&array[cont]);
		cont++;

	}

}

// =-=-=-=-=-=-=-=-=-=-=-=-=-=-= //

int main(void)
{

	printf("Selecione o algoritmo que você quer executar:  \n");
	printf("1) Insertion Sort                              \n");
	printf("2) Selection Sort                              \n");
	printf("3) Bubble Sort                                 \n");
	printf("4) Heap Sort                                   \n");
	printf("5) Merge Sort                                   \n");
	printf("6) Quick Sort                                   \n");
	
	int alg;
	scanf("%d", &alg);

	if (alg == 1) {

		clock_t inicio, fim;
		inicio = clock();

		int tamanho = 0;
		int vetor[100000];
		selecionar_arquivo(vetor, &tamanho);
		insertion(vetor, tamanho);
		imprimir(vetor, tamanho);
		
		fim = clock();
		double Tempo = (fim - inicio) * 1000.0 / CLOCKS_PER_SEC;
		
		printf("\n\nTempo de execução: %g ms \n", Tempo);

	} else if (alg == 2) {

		clock_t inicio, fim;
		inicio = clock();

		int tamanho = 0;
		int vetor[100000];
		selecionar_arquivo(vetor, &tamanho);
		selection(vetor,tamanho);
		imprimir(vetor,tamanho);
		
		fim = clock();
		double Tempo = (fim - inicio) * 1000.0 / CLOCKS_PER_SEC;

		printf("\n\nTempo de execução: %g ms \n", Tempo);

	} else if (alg == 3) {

		clock_t inicio, fim;
		inicio = clock();

		int tamanho = 0;
		int vetor[100000];
		selecionar_arquivo(vetor, &tamanho);
		bubble(vetor,tamanho);
		imprimir(vetor,tamanho);
		
		fim = clock();
		double Tempo = (fim - inicio) * 1000.0 / CLOCKS_PER_SEC;

		printf("\n\nTempo de execução: %g ms \n", Tempo);

	} else if (alg == 4) {

		clock_t inicio, fim;
		inicio = clock();

		int tamanho = 0;
		int vetor[100000];
		selecionar_arquivo(vetor, &tamanho);
		heapSort(vetor, tamanho);
		imprimir(vetor, tamanho);
		
		fim = clock();
		double Tempo = (fim - inicio) * 1000.0 / CLOCKS_PER_SEC;
		
		printf("\n\nTempo de execução: %g ms \n", Tempo);

	} else if (alg == 5) {

		clock_t inicio, fim;
		inicio = clock();

		int tamanho = 0;
		int vetor[100000];
		selecionar_arquivo(vetor, &tamanho);
		mergeSort(vetor, tamanho);
		imprimir(vetor, tamanho);
		
		fim = clock();
		double Tempo = (fim - inicio) * 1000.0 / CLOCKS_PER_SEC;
		
		printf("\n\nTempo de execução: %g ms \n", Tempo);    
	} else if (alg == 6) {

		clock_t inicio, fim;
		inicio = clock();

		int tamanho = 0;
		int vetor[100000];
		selecionar_arquivo(vetor, &tamanho);
		quickSort(vetor, tamanho);
		imprimir(vetor, tamanho);

		fim = clock();
		double Tempo = (fim - inicio) * 1000.0 / CLOCKS_PER_SEC;
		
		printf("\n\nTempo de execução: %g ms \n", Tempo);    
	} else {
		
		printf("Valor inválido!");

	}
	

}

File: sorting_algorithms.hs
{-# LANGUAGE BangPatterns #-}
import Control.DeepSeq
import Text.Printf
import System.IO
import System.CPUTime
import Control.Exception
import Data.Time.Clock

main :: IO()
main = do
    
    putStrLn "Selecione o algoritmo que você quer executar:"
    putStrLn "1) Insertion Sort"
    putStrLn "2) Selection Sort"
    putStrLn "3) Bubble Sort"
    putStrLn "4) Heap Sort"
    putStrLn "5) Merge Sort"
    putStrLn "6) Quick Sort"

    algoritmo <- getLine
    let escolha | algoritmo == "1" = insertion
                | algoritmo == "2" = selection
                | algoritmo == "3" = bubble
                | algoritmo == "4" = heap
                | algoritmo == "5" = merge
                | algoritmo == "6" = quick

    putStrLn "Selecione o algoritmo que você quer executar:"
    putStrLn "1) 1000"
    putStrLn "2) 5000"
    putStrLn "3) 10000"
    putStrLn "4) 20000"
    putStrLn "5) 50000"
    putStrLn "6) 75000"
    putStrLn "7) 100000"

    arq <- getLine
    let arquivo | arq == "1" = "1000.txt"
                | arq == "2" = "5000.txt"
                | arq == "3" = "10000.txt"
                | arq == "4" = "20000.txt"
                | arq == "5" = "50000.txt"
                | arq == "6" = "75000.txt"
                | arq == "7" = "100000.txt"
    conteudo <- readFile arquivo
    let qtd_numeros = map (\x -> read x::Int) (words conteudo)

    inicio <- getCPUTime
    let !sortMap = rnf $ escolha qtd_numeros
    fim <- getCPUTime
    let tempo = fromIntegral(fim - inicio) / (10 ^ 12)
    printf "Tempo de execução: %0.5f ms\n" (tempo :: Double)


-- Insertion Sort

insertion ::(Ord a) => [a] -> [a]
insertion [] = []
insertion (x:xs) = ordenar x (insertion xs)

ordenar::(Ord a) => a -> [a] -> [a]
ordenar r [] = [r]
ordenar r (x:xs) | r <= x = r:x:xs
                 | otherwise = x : ordenar r xs  

-- Selection Sort

menor :: (Ord a) => [a] -> a
menor [r] = r
menor (r:x:xs) | r <= x = menor(r:xs)
               | otherwise = menor(x:xs)


selection :: (Ord a) => [a] -> [a]
selection [] = []
selection xs = let  x = menor xs  
           in  x : selection (excluir x xs) 


excluir :: (Ord a) => a -> [a] -> [a]
excluir _ [] = []
excluir r (x:xs)    | r == x = xs
                    | otherwise = x : excluir  r xs

-- Bubble Sort

bubble :: (Ord a) => [a] -> [a]
bubble [] = []
bubble [r] = [r]
bubble (r:x:xs)    | ordenado par = par 
                   | otherwise =  bubble par
    where par = menor' r x : bubble (maior r x : xs) 
 
ordenado :: (Ord a) => [a] -> Bool
ordenado [] = True
ordenado [r] = True
ordenado (r:x:xs) |  r <= x = ordenado (x:xs)
                  | otherwise =  False  


menor' :: (Ord a) => a -> a -> a
menor' r x  | r <= x = r
           | otherwise = x

maior :: (Ord a) => a -> a -> a
maior r x  | r >= x = r
           | otherwise = x

-- Heap Sort

trocar :: [Int] -> Int -> Int -> [Int]
trocar x a b | a == b = x
             | b < a = trocar x b a
             | otherwise = take a x ++ [x !! b] ++ drop (a + 1) (take b x) ++ [x !! a] ++ drop (b + 1) x

excluirUltimo :: [Int] -> [Int]
excluirUltimo [x] = []
excluirUltimo [] = []
excluirUltimo x = take (length x-1) x


ordenarArvore :: [Int] -> Int -> [Int]
ordenarArvore x i | direita i < length x && (x !! direita i >= x !! esquerda i) && (x !! direita i > x !! i) = ordenarArvore (trocar x i (direita i)) (direita i)
                  | esquerda i < length x && (x !! esquerda i > x !! i) = ordenarArvore (trocar x i (esquerda i)) (esquerda i)
                  | otherwise = x
                    where esquerda t = 2 * t + 1
                          direita t = 2 * t + 2

maxHeapify :: [Int] -> [Int]
maxHeapify [x] = [x]
maxHeapify [] = []
maxHeapify (x:xs) = maxHeapify (ordenarArvore (last xs : excluirUltimo xs) 0) ++ [x]

buildMaxHeap :: [Int] -> Int -> [Int]
buildMaxHeap x y | y >= 0 = buildMaxHeap (ordenarArvore x y) (y - 1)
                 | otherwise = x


heap :: [Int] -> [Int]
heap x = maxHeapify (buildMaxHeap x ((length x-2) `div` 2))

-- Merge Sort

intercala :: Ord a => [a] -> [a] -> [a]
intercala  x [ ] = x
intercala [ ] y  = y
intercala (x:xs) (y:ys) | x <= y    = x : intercala xs (y:ys)
                        | otherwise = y : intercala (x:xs) ys

inicio :: [a] -> [a]
inicio xs = take (length xs `div` 2) xs

fim :: [a] -> [a]
fim xs = drop (length xs `div` 2) xs
                        
merge :: Ord a => [a] -> [a]
merge [ ] = [ ]
merge [x] = [x]
merge xs  = intercala (merge (inicio xs)) (merge (fim xs))

                  
-- Quick Sort


left :: (Ord a) => [a] -> [a]
left [] = []
left (x:xs) = filter (<x) xs

right :: (Ord a) => [a] -> [a]
right [] = []
right (x:xs) = filter (>=x) xs

quick :: (Ord a) => [a] -> [a]
quick [] = []
quick (x:xs) = quick (left (x:xs)) ++ [x] ++ quick (right (x:xs))
